truffleruby/src/main/java/org/truffleruby/core/rubinius/IONodes.java

```diff
@@ -799,7 +801,9 @@ public int accept(DynamicObject io) {
             final int newFd;
 
             try {
-                newFd = ensureSuccessful(nativeSockets().accept(fd, memoryManager().newPointer(address), addressLength));
+                newFd = getContext().getThreadManager().runUntilResult(this,
+                        () -> ensureSuccessful(nativeSockets().accept(fd, memoryManager().newPointer(address), addressLength)),
+                        () -> nativeSockets().shutdown(fd, SHUT_RD));
```

Comment:
Thanks for looking into this more. I was hoping `shutdown` would be friendlier.

We could rewrite the code to use `select`, but I don't think implementing `accept` in terms of `select` is what we want to do. It's supposed to be a straightforward POSIX call mapping. MRI also has `accept_nonblock`, which is essentially what you're proposing for the first case. We may just have to make the `fcntl` call. I think a similar solution is unavoidable for `read`.